[Feat] Security 설정 및 Member 엔티티 골격 구성

asd42270 opened the pull request:
# ⭐️[Feat] Security 설정 및 Member 엔티티 골격 구성
- 기획 전 진행 가능한 Security 설정에 대한 PR입니다.

## 🛠️ 변경 사항
**배경 및 목적 :**
- 기획 이후 개발 소요시간 단축을 위해 로그인 기능을 미리 작성
- FE 협의 내용에 따라 토큰 로그인으로 헤더에 삽입하는 방향으로 구성

**변경 사항:**
- 의존성 추가
  - MySQL Connector
  - JJWT

- Config
  - Security Enable 설정을 위한 Config 파일 추가
  - BCryptPasswordEncoder 채택
  - csrf, cors, httpBasic 비활성화

- Token Generator
  - key, 토큰 유효기간 하드 코딩으로 주입 - 추후 yml 업로드시 yml을 통한 주입으로 변경 예정

- UserDetails
  - MemberPrincipal 네이밍 채택

- Filter
  - JwtAuthenticationFilter 구현

- Exception
  - 403 계열 에러코드 추가
  - JwtNotFoundException, MemberDuplicatedException, MemberNotExistException 생성

- Member
  - 기획 전 작성이므로 email,password 컬럼 사용
  - email 기반 조회가 빈번할 것으로 예상되어 컬럼으로 인덱싱
  - Controller 반환 타입 미정으로 인한 Void 리턴 - FE와 회의 후 변경 예정
  - Repository에 default 키워드 사용으로 중복될 예외 처리 공토오하
  - dto 내부에 request,response를 구분하는 패키지 구성 채택
  - dto에 record 타입 사용


## 관련 이슈

- #22 
- #29 

## 개발 유형

- [ ] 버그 수정
- [x] 새로운 기능 개발
- [ ] 리팩토링
- [ ] 테스트 코드 작성
- [ ] 문서 업데이트
- 기타 (아래에 설명 추가)

## 작업 기간

- 작업 시작일: (2024-05-01)
- 작업 종료일: (2024-05-01)

## 유의사항

- Controller 레이어에서 User정보 필요한 경우 @AuthenticationPrincipal 어노테이션 사용으로 Member 반환 가능

## 참고문헌

-

asd42270 committed:
[feat] SecurityConfig 클래스 추가
- PasswordEncoder 추가: 빈 주입, BCrypt 암호화 알고리즘 방식 사용
- Security 빈 추가: csrf,httpBasic,cors disable 처리, api 만들기 전 모든 요청 허용(permitAll), session 비활성화

asd42270 committed:
[feat] Jwt 사용을 위한 의존성 추가

asd42270 committed:
[feat] Member 엔티티 추가
- Member: 로그인을 위해 필수 컬럼 생성
- MemberRepository: loadUserByUsername을 생성하여 repository 내부에서 UserException 다루도록 조치
- UserException: member 패키지 내부에 UserException 추가
- ErrorCode: 403에러 MEMBER_NOT_FOUND 에러 추가

asd42270 committed:
[feat] Jwt 설정 파일 추가
- MemberRepository: 사용하지 않는 import문 삭제
- MemberPrincipal: Security 로그인 설정을 위해 UserDetails를 구현한 구현체
- MemberPrincipalService: Member를 MemberPrincipal로 매핑

asd42270 committed:
[feat] Jwt 설정 파일을 추가한다.
- ErrorCode: 토큰 관련 에러 추가
- JwtAuthenticationFilter: Jwt 토큰 필터 설정
- JwtNotFoundException: 토큰 에러 추가
- JwtTokenGenerator: 토큰 생성을 위한 클래스

asd42270 committed:
[feat] Jwt 설정 파일을 추가한다.
- Waring 사항 조치

asd42270 committed:
[feat] Jwt, Security 설정
- JwtAuthenticationFilter: accessToken이 null일 때 throw를 하는 것이 아니라 바로 다음 필터로 이동하로고 수정한다.
- JwtTokenGenerator: 토큰 접근 만료 시간 임시 설정, key 더 길게 변경
- SecurityConfig: jwt인증필터를 거치도록 수정

asd42270 committed:
[feat] Member 도메인 임시 작성
- Member: 패키지 위치 변경, 빌더 어노테이션 사용(기본 생성자, 생성자 어노테이션 사용-접근권한 protected로 설정)
- MemberRepository: 이메일 중복 여부 확인을 위해 existsByEmail 로직 작성
- MemberService: register로직 작성- 요청 데이터의 유효성을 검사하고, 인코딩 후 Member로 바꾸어 DB에 저장
login 로직 - 요청 데이터의 비밀버호를 DB 데이터와 매치시켜 일치하면 토큰을 전달
passwordMatcher- password 일치여부 확인
- MemberController: 회원가입과 로그인의 컨트롤러 로직을 작성
- MemberDuplicatedException: 중복 이메일에 대한 예외 작성
- MemberNotExistException: 존재하지 않는 이메일에 대한 예외 작성
- MemberPrincipal: Member 클래스 위치 변경으로 인한 커밋
- MemberRegisterRequestDto: 회원가입 요청 시 받을 Dto 생성
- MemberRequestDto : 로그인 요청 시 받을 Dto 생성

- Postman으로 로그인 시 헤더에 토큰값 전달 되는 것 확인
- ErrorCode: DUPLICATED_EMAIL 추가

asd42270 committed:
[Fix] 사용하지 않는 import문 삭제

Galmaeki commented:
Pr 문서화 완료했습니다
이슈 문서화 예정이며 부족한 부분 지적해주세요!

ghrltjdtprbs commented on src/main/java/com/final_10aeat/domain/member/controller/MemberController.java:
```diff
@@ -0,0 +1,35 @@
+public class MemberController {
+    private final MemberService memberService;
+
+    @PostMapping("/register")
```
URI를 동작이 아닌 자원을 지칭하도록 변경하면 좀 더  RESTful 한 설계가 될 것 같습니다!

ghrltjdtprbs commented on src/main/java/com/final_10aeat/domain/member/dto/request/MemberRegisterRequestDto.java:
```diff
@@ -0,0 +1,6 @@
+package com.final_10aeat.domain.member.dto.request;
+
+public record MemberRegisterRequestDto(
```
DTO에 record를 사용하셨네요! record 를 사용하시면 코드가 간결해지고 데이터의 불변성을 보장하니 좋은 선택인 것 같습니다 ㅎㅎ 👍👍
아시겠지만 record 클래스에도  @Email, @NotBlank, @Size  같은 Hibernate Validator 사용이 가능하니 유효성 검사 로직을 추가해 보시는건 어떨까요?

ghrltjdtprbs commented on src/main/java/com/final_10aeat/domain/member/dto/request/MemberRequestDto.java:
```diff
@@ -0,0 +1,8 @@
+   String email,
+   String password
+) {
+
```
마찬가지로 유효성 검사 로직을 추가하면 좋을 것 같습니다.😊 
그리고 현재 DTO 클래스들의 이름이 의미를 잘 전달하고 있지만, 두 개의 DTO 의 차이를 좀 더 명확하게 하기 위해LoginRequestDto, RegisterRequestDto 등으로 변경하시는 것은 어떨까요?

ghrltjdtprbs commented on src/main/java/com/final_10aeat/domain/member/entity/Member.java:
```diff
@@ -0,0 +1,27 @@
+    @Column
+    private String password;
+
+}
```
코드 스타일 통일하는건 어떨까요?

1. 메서드 선언의 맨 위와 아래는 줄바꿈 X
2. 클래스 선언의 맨 위는 줄바꿈 O, 맨 아래는 줄바꿈 X
3. 파일의 맨 아래에 줄바꿈 삽입 

맨 아래 개행이 필요한 이유: https://seongwon.dev/Git/20220303-%ED%8C%8C%EC%9D%BC%EC%9D%98_%EB%A7%88%EC%A7%80%EB%A7%89_%EA%B0%9C%ED%96%89/

ghrltjdtprbs commented on src/main/java/com/final_10aeat/domain/member/exception/MemberNotExistException.java:
```diff
@@ -0,0 +1,14 @@
+    public MemberNotExistException() {
+        super(ERROR_CODE);
+    }
+
```
코드스타일 관련 위 리뷰와 동일합니다.😊

ghrltjdtprbs commented on src/main/java/com/final_10aeat/domain/member/repository/MemberRepository.java:
```diff
@@ -0,0 +1,17 @@
+package com.final_10aeat.domain.member.repository;
+
+
```
개행에 신경써주세요! 커밋 전 줄정렬을 해주거나 한 번 더 체크하면 더 깔끔한 코드가 됩니다 ㅎㅎ👍

ghrltjdtprbs commented on src/main/java/com/final_10aeat/domain/member/repository/MemberRepository.java:
```diff
@@ -0,0 +1,17 @@
+    Boolean existsByEmail(String email);
+    Optional<Member> findByEmail(String email);
+
+    default Member loadUserByUsername(String email){
```
loadUserByUsername 메서드가 현재 인터페이스 내에 정의되어 있지만, 프로젝트의 전체 구조를 고려하여 서비스 계층 또는 유틸리티 클래스로 분리하는 것도 고려해 볼 수 있을 것 같습니다! 
이는 인터페이스의 선언적 역할을 유지하면서도, 메서드의 재사용성을 높일 수 있습니다.😊

ghrltjdtprbs commented on src/main/java/com/final_10aeat/global/config/SecurityConfig.java:
서비스 계층을 더 세분화하여 AuthenticationService와 같은 별도의 서비스로 분리하는 것을 고려해 보시는 건 어떠신가요? 이는 전체 서비스 계층의 역할을 더 명확하게 하고, 코드의 유지보수성을 높일 수 있습니다! 하지만 그냥 제안이니 이대로 진행하셔도 상관없습니다! 👍👍
+개행도...

ghrltjdtprbs commented on src/main/java/com/final_10aeat/domain/member/service/MemberService.java:
```diff
@@ -0,0 +1,57 @@
+
+@Service
+@RequiredArgsConstructor
+public class MemberService {
```
클래스 레벨에 트랜잭션을 적용하면 좋을 것 같습니다!
아시겠지만, 트랜젝션을 적용하면 데이터베이스에 대한 일련의 작업을 하나의 단위로 묶어 처리하고 ACID 원칙을 보장한다는 장점이 있습니다 ㅎㅎ 
또한 읽기 전용 트랜젝션으로 성능 향상도 노려볼 수 있습니다 👍

ghrltjdtprbs commented on src/main/java/com/final_10aeat/global/security/principal/MemberPrincipalService.java:
```diff
@@ -0,0 +1,19 @@
+@Component //Service보다는 Component의 성격에 가까움.
+@RequiredArgsConstructor
+//TODO: 클래스 이름 변경 고려
+public class MemberPrincipalService {
```
클래스 이름이 MemberPrincipalService인데, 이는 서비스 계층과의 혼동이 있을 수 있습니다. 클래스의 목적과 기능을 반영한 이름으로 변경하는 것이 좋습니다. 
예를 들어, MemberDetailsProvider 또는 UserDetailsProvider와 같은 이름은 어떨까요?

ghrltjdtprbs commented on src/main/java/com/final_10aeat/global/security/principal/MemberPrincipalService.java:
```diff
@@ -0,0 +1,19 @@
+
+
+
+    public UserDetails loadUserByUsername(String email){
```
loadUserByUsername 메서드가 스프링 시큐리티의 UserDetailsService 인터페이스의 메서드와 동일한 기능을 수행하는 것 같네요! 
이 부분을 명시적으로 구현하여 클래스의 구조를 명확히 하고, 스프링 시큐리티와의 통합을 강화하는건 어떨까요? 
예를 들어, MemberPrincipalService 클래스가 UserDetailsService 인터페이스를 직접 구현하고, 해당 메서드를 재정의하도록 변경할 수 있습니다.😊
```java
@Override
    public UserDetails loadUserByUsername(String email) {
        return new MemberPrincipal(memberRepository.loadUserByUsername(email));
    }
```
이렇게 하면 클래스가 스프링 시큐리티의 구조에 직접 통합되며, 인증 및 권한 부여 관련 로직을 더 쉽게 유지 관리할 수 있습니다.👍👍

ghrltjdtprbs commented on src/main/java/com/final_10aeat/global/exception/ErrorCode.java:
제가 ErrorCode 사용법을 제대로 알려드리지 못한 것 같습니다! 
프로젝트에서 다루는 예외가 대부분 클라이언트 에러이므로, HTTP 상태 코드 기반의 분류는 추적이 어렵습니다. 그렇기 때문에 도메인 중심으로 분류하는 것을 추천드려요! ( //4xx -> // MEMBER)

ghrltjdtprbs commented on src/main/java/com/final_10aeat/global/exception/ErrorCode.java:
```diff
@@ -6,9 +6,15 @@
 public enum ErrorCode {
 
+    //4xx
+    MEMBER_NOT_EXIST(HttpStatus.FORBIDDEN, "존재하지 않는 사용자입니다."),
```
**MEMBER_NOT_EXIST :** 
Forbidden은 주로 사용자에게 해당 리소스에 접근할 권한이 없을 때 사용됩니다. 그러나 "존재하지 않는 사용자" 상황에서는 404 Not Found가 더 적절할 수 있습니다. 이는 해당 사용자가 존재하지 않아 리소스를 찾을 수 없다는 의미로 해석될 수 있습니다.

**TOKEN_NOT_FOUND :** 
Unauthorized는 토큰을 통해 사용자 인증이 필요하나, 토큰이 없거나 잘못된 경우 사용되는 코드입니다. 그러므로 "토큰을 찾을 수 없다"는 상황에서는 Unauthorized가 더 적절할 것 같습니다! 😊

ghrltjdtprbs reviewed (changes requested):
수고하셨습니다! 코드 보고 몇 가지 개선 사항을 남겨뒀으니 참고해주세요! 👍👍

ghrltjdtprbs commented:
추가로 pr 컨벤션과 브랜치 컨벤션에 대해 리뷰 남깁니다! 
pr 컨벤션은 커밋 컨벤션과 동일하게 이슈 번호 필요없이 [Feat] 만 붙여주시면 됩니다! 😊
브랜치는 " feature/이슈 번호" 입니다! 참고로 hotfix는 이슈번호 필요없이  hotfix로만 사용하시면 됩니다! 
(feat#22 -> feature/22)

Backend Convention : https://github.com/orgs/10aeat/discussions/3

ghrltjdtprbs commented on src/main/java/com/final_10aeat/domain/member/entity/Member.java:
```diff
@@ -0,0 +1,27 @@
+))
+@AllArgsConstructor(access = AccessLevel.PROTECTED)
+@NoArgsConstructor(access = AccessLevel.PROTECTED)
+public class Member extends BaseTimeEntity {
```
회원 정보는 무조건 soft delete기능이 들어갑니다! SoftDeletableBaseTimeEntity 를 이용해주세요 ㅎㅎ 😊

asd42270 committed:
[feat] application.yml 기본 설정
- 로컬 db url, hibernate 설정

asd42270 committed:
[feat] ErrorCode
- 주석 4xx에러-> Member 도메인으로 변경
- MEMBER_NOT_EXIST: FORBIDDEN->BADREQUEST로 변경
- TOKEN_NOT_FOUND: FORBIDDEN-> UNAUTHORIZED으로 변경

asd42270 committed:
[fix] Jwt 필터 수정
- MemberPrincipalService-> MemberDetailsProvider 클래스 이름 변경으로 인하여 클래스 이름 수정
- 개행 구조 수정

asd42270 committed:
[fix] MemeberEntity 수정
- BaseTimeEntity-> SoftDeletableBaseTimeEntity 상속으로 변경
- Index 테이블 수정: coulmnList에 deletedAt 컬럼 추가

asd42270 committed:
[fix] MemberController 수정
- URI 수정: Restful하게 변경

asd42270 committed:
[fix] MemberDetailsProvider 수정
- MemberPrincipalService -> MemberDetailsProvider로 클래스 이름 변경
- UserDetailService를 상속받아 해당 인터페이스의 loadUserByUsername를 구현하도록 변경
- MemberRepository의 loadUserByUsername로 인한 orElseThrow추가

asd42270 committed:
[fix] MemberLoginRequestDto
- MemberRequestDto: MemberLoginRequestDto로 클래스 이름 변경

asd42270 committed:
[fix] MemberNotExistException
- 코드 스타일 수정

asd42270 committed:
[fix] MemberRepository 수정
- loadUserByUsername 메소드 삭제

asd42270 committed:
[fix] MemberService 수정
- @Transactional 추가
- login메서드에 @Transactional(readOnly = true) 추가
- MemberRequestDto-> MemberLoginRequestDto로 변경되어 수정

asd42270 committed:
[fix] SecurityConfig 수정
- 코드 스타일 수정

asd42270 committed:
[fix] MemberRepository 수정
- 사용하지 않는 import문 삭제

asd42270 committed:
Merge branch '10aeat:develop' into feat#22

Galmaeki commented on src/main/java/com/final_10aeat/domain/member/controller/MemberController.java:
```diff
@@ -0,0 +1,35 @@
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequiredArgsConstructor
```
```java
@RequestMaping("/members")
```
중복되는 경로는 묶어줘도 좋을 것 같습니다

asd42270 replied:
수정했습니다~~~

ghrltjdtprbs reviewed (approved):
변경사항 확인했습니다! 수고하셨어요ㅎㅎ 👍👍

Galmaeki reviewed (changes requested):
리뷰사항 남겨놨습니다!
확인 부탁드려요!

asd42270 committed:
[fix] MemberController 수정
- @Requestmapping으로 mapping의 공통부분을 묶어준다.

asd42270 committed:
Merge remote-tracking branch 'origin/feat#22' into feat#22

Galmaeki reviewed (approved):
수고하셨습니다
